Port over doc comments in route macros.

erikjohnston opened the pull request:
## PR Type
<!-- What kind of change does this PR make? -->
<!-- Bug Fix / Feature / Refactor / Code Style / Other -->
Bug Fix / Feature

## PR Checklist
<!-- Check your PR fulfills the following items. ->>
<!-- For draft PRs check the boxes as you complete them. -->

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.


## Overview

This allows documentation on the functions to appear in the generated
docs, e.g.:

```rust
/// The index page
async fn index() -> &'static str {
    "Hello!"
}
```

Closes #1389

--- 

I'm afraid I have no idea how to add tests for this beyond making sure this doesn't break when using docstrings.

erikjohnston committed:
Port over doc comments in route macros.

This allows documentation on the functions to appear in the generated
docs, e.g.:

```rust
/// The index page
async fn index() -> &'static str {
    "Hello!"
}
```

robjtede commented:
Does this handle multi line doc-comments? I don't recall if they're converted to one doc attribute with line breaks or multiple doc attrs.

erikjohnston committed:
Handle multi-line doc comments.

These are converted into multiple attributes, and so we need to store
them all.

erikjohnston commented:
> Does this handle multi line doc-comments? I don't recall if they're converted to one doc attribute with lie breaks or multiple doc attrs.

Huh, you're right, had no idea that is multi line comments worked. I've pushed a fix and manually tested.

Tangentially, I am wondering whether it'd be better to just copy all attributes over? It's probably not *that* useful, but would mean e.g. that people could have manually added `#[allow(missing_docs)]` to the structs.

robjtede committed:
Update CHANGES.md

robjtede committed:
Merge branch 'master' into codegen_port_doc_comments

robjtede commented:
> Tangentially, I am wondering whether it'd be better to just copy all attributes over?

I'm apprehensive about that because it may seem strange that attributes meant for functions will not work on the generated struct. If you don't feel strongly that it should be done I'd like to wait until it is raised specifically.

jplatte commented on actix-web-codegen/src/route.rs:
```diff
@@ -221,6 +224,20 @@ impl Route {
+        let mut doc_attribute = Vec::new();
+        for attr in &ast.attrs {
+            if let Some(ident) = attr.path.get_ident() {
+                if ident == "doc" {
```
```suggestion
            if attr.path.get_ident().is_ident("doc") {
```

robjtede replied:
https://docs.rs/syn/1.0.60/syn/struct.Path.html#method.is_ident

jplatte commented on actix-web-codegen/src/route.rs:
```diff
@@ -221,6 +224,20 @@ impl Route {
+        //
+        // Note that multi line doc comments are converted to multiple doc
+        // attributes.
+        let mut doc_attribute = Vec::new();
```
```suggestion
        let mut doc_attributes = Vec::new();
```

then the return value can also use struct init shorthand for this.

jplatte replied:
Actually I think this + the whole loop can be simplified to

```rust
let doc_attributes = ast.attrs.iter().filter(|attr| attr.path.is_ident("doc")).cloned().collect();
```

jplatte commented on actix-web-codegen/src/route.rs:
```diff
@@ -265,6 +283,7 @@ impl ToTokens for Route {
                     methods,
                 },
             resource_type,
+            doc_attributes: doc_attribute,
```
```suggestion
            doc_attributes,
```

jplatte commented on actix-web-codegen/src/route.rs:
```diff
@@ -286,7 +305,13 @@ impl ToTokens for Route {
+            .iter()
+            .map(ToTokens::to_token_stream)
+            .collect();
+
```
This is not needed, you can just use a repetition in the output. Suggestion in next review comment.

jplatte commented on actix-web-codegen/src/route.rs:
```diff
@@ -286,7 +305,13 @@ impl ToTokens for Route {
+            .collect();
+
         let stream = quote! {
+            #doc_comment_token_stream
```
```suggestion
            #(#doc_comments)*
```

erikjohnston committed:
Apply suggestions from code review

Co-authored-by: Jonas Platte <jplatte@users.noreply.github.com>

erikjohnston committed:
Simplify creating and using doc_attributes vec

erikjohnston commented:
Thank you both, my proc macro foo is still quite weak :heart:
